File: libjs/wModel/button.js
"use strict";

var Model = require("./model");
var ModelString = require("./string");

var Vocabulary = require("../wType/vocabulary");
var Boolean = require("../wType/boolean");
var Address = require("../wType/address");
var String = require("../wType/string");

var Button = Model.inherit({
    "className": "Button",
    "constructor": function(address) {
        Model.fn.constructor.call(this, address);
        
        this._enabled = true;
        this._hasImage = false;
        this._hasLabel =false;
        this._imageName = undefined;
        this._label = undefined;
        
        this.addHandler("get");
        this.addHandler("activate");
    },
    "setImage": function(name) {
        if (this._hasImage) {
            if (this._imageName !== name) {
                this._image = name;
                var vc = new Vocabulary();
                vc.insert("image", new String(this._imageName));
                this.broadcast(vc, "changeImage");
            }
        } else {
            this._image = name;
            this._hasImage = true;
            var vc = new Vocabulary();
            vc.insert("image", new String(this._imageName));
            this.broadcast(vc, "setImage");
        }
    },
    "setEnabled": function(enabled) {
        if (enabled !== this._enabled) {
            this._enabled = enabled;
            var vc = new Vocabulary();
            vc.insert("enabled", new Boolean(this._enabled));
            this.broadcast(vc, "setEnabled");
        }
    },
    "setLabel": function(text) {
        if (this._hasLabel) {
            this._label.set(text);
        } else {
            this._label = new ModelString(this._address["+"](labelHop), text);
            this.addModel(this._label);
            var vc = new Vocabulary();
            vc.insert("hasLabel", new Boolean(true));
            vc.insert("label", this._label.getAddress());
            this.broadcast(vc, "setLabel");
            this._hasLabel = true;
        }
    },
    "_h_subscribe": function(ev) {
        Model.fn._h_subscribe.call(this, ev);
        
        this._h_get(ev);
    },
    "_h_get": function(ev) {
        var vc = new Vocabulary();
        vc.insert("hasImage", new Boolean(this._hasImage));
        if (this._hasImage) {
            vc.insert("image", new String(this._imageName));
        }
        vc.insert("hasLabel", new Boolean(this._hasLabel));
        if (this._hasLabel) {
            vc.insert("label", this._label.getAddress());
        }
        vc.insert("enabled", new Boolean(this._enabled));
        
        this.response(vc, "get", ev);
    },
    "_h_activate": function() {
        if (this._enabled) {
            this.trigger("activated");
        }
    }
});

var labelHop = new Address(["label"]);

module.exports = Button;

File: lib/wModel/button.cpp
#include "button.h"

M::Button::Button(const W::Address& address, QObject* parent):
    M::Model(address, parent),
    enabled(true),
    hasImage(false),
    hasLabel(false),
    imageName(0),
    label(0)
{
    W::Handler* get = W::Handler::create(address + W::Address({u"get"}), this, &M::Button::_h_get);
    W::Handler* activate = W::Handler::create(address + W::Address({u"activate"}), this, &M::Button::_h_activate);
    
    addHandler(get);
    addHandler(activate);
}

M::Button::~Button()
{
    if (hasImage) {
        delete imageName;
    }
}

void M::Button::setImage(const W::String& p_image)
{
    
    if (hasImage) {
        if (*imageName != p_image) {
            imageName = static_cast<W::String*>(p_image.copy());
            W::Vocabulary* vc = new W::Vocabulary();
            vc->insert(u"image", p_image);
            broadcast(vc, W::Address{u"changeImage"});
        }
    } else {
        imageName = static_cast<W::String*>(p_image.copy());
        hasImage = true;
        W::Vocabulary* vc = new W::Vocabulary();
        vc->insert(u"image", p_image);
        broadcast(vc, W::Address{u"setImage"});
    }
    hasImage = true;
}

void M::Button::setEnabled(bool p_enabled)
{
    if (enabled != p_enabled) {
        enabled = p_enabled;
        W::Vocabulary* vc = new W::Vocabulary();
        vc->insert(u"enable", new W::Boolean(enabled));
        broadcast(vc, W::Address{u"setEnabled"});
    }
}

void M::Button::setLabel(const W::String& p_label)
{
    if (hasLabel) {
        label->set(p_label);
    } else {
        label = new M::String(p_label, address + W::Address{u"label"});
        addModel(label);
        W::Vocabulary* vc = new W::Vocabulary();
        vc->insert(u"hasLabel", new W::Boolean(true));
        vc->insert(u"label", label->getAddress());
        broadcast(vc, W::Address{u"setLabel"});
        hasLabel = true;
    }
}

void M::Button::h_subscribe(const W::Event& ev)
{
    M::Model::h_subscribe(ev);
    
    h_get(ev);
}

void M::Button::h_get(const W::Event& ev)
{
    W::Vocabulary* vc = new W::Vocabulary();
    vc->insert(u"hasImage", new W::Boolean(hasImage));
    if (hasImage) {
        vc->insert(u"image", imageName->copy());
    }
    vc->insert(u"hasLabel", new W::Boolean(hasLabel));
    if (hasLabel) {
        vc->insert(u"label", label->getAddress());
    }
    vc->insert(u"enabled", new W::Boolean(enabled));
    
    response(vc, W::Address({u"get"}), ev);
}

void M::Button::h_activate(const W::Event& ev)
{
    if (enabled) {
        emit activated();
    }
}

M::Model::ModelType M::Button::getType() const
{
    return M::Model::button;
}

void M::Button::set(const W::Object& value)
{
    throw 14;   //what do you expect here? not implemented, and not sure it ever would be
}

void M::Button::set(W::Object* value)
{
    set(*value);
}
